### ISSUE TITLE
Class not found if the class line ends with a comment

### ISSUE THREAD

**Issue by Taitava** [silverstripe/silverstripe-framework#6615]
SilverStripe version: 3.4.1

I run into this issue when I added a comment to one of my Controller classes which I knew was previously working well. I haven't tested this with other kind of classes (i.e. DataObjects), but I think it would be reasonable if the same issue affects other classes too.

1. Create a new Controller class and add a single line comment:
```PHP
class MyController extends Controller //Some comment
{
    private static $allowed_actions = array(
        'test' => true,
    );
	
    private static $url_handlers = array(
        'test'	=> 'test',
    );

    public function test()
    {
        return 'MyController works.';
    }
}
```

2. Define the route:
```YAML
Director:
  rules:
    'MyController//$Action': 'MyController'
```

3. Run `/dev/build?flush=all`

4. Run `/MyController/test`. You should now get this error: `[User Error] Uncaught ReflectionException: Class MyController does not exist`.

(A couple of side notes: I usually name my controllers like `Something_Controller` but in this example I didn't use that pattern. Also it's not a huge issue being not able to put comments in the same line with the `class` keyword, but I just wanted to open this issue so that if someone else is wondering why their class is not found, then perhaps this issue can be of help. The reason why I originally put the comment to that line was that I wanted to write a note about why I was extending another custom controller instead of straight the base Controller class, so I thought it would be reasonable to put the comment near the `extends` keyword. Well, I can place it elsewhere though :).)

**Comment by gregsmirnov** [comment]
Why don't you use PHPDoc for this purpose?

**Comment by Taitava** [comment]
Well, that's what I started to use after I found out that I can't put a comment at the same line with the `class` definition.

But still, I think we have a bug somewhere. It's confusing not to be able to put a comment in a place where PHP's syntax clearly allows comments to exist :). Definitively it's nothing big and should have a low priority but would be good to check at some point.

**Comment by dhensby** [comment]
This definitely sounds like a bug, I'm looking into it

**Comment by dhensby** [comment]
OK - so the problem is you can't put comments anywhere in the declaration line

**Comment by dhensby** [comment]
@Taitava I've opened a PR with a fix - would you mind testing it for me?

**Comment by Taitava** [comment]
@dhensby Sure, just a moment :).

**Comment by Taitava** [comment]
@dhensby Unfortunately I still get the same error. I'm not good with git (although I do use it almost daily) so I downloaded the whole framework from your `pulls/3.4/fix-class-manifest` branch and copied it over my project's framework folder replacing everything. I checked that the changes you made in your commit existed in my project's files, so I'm confused why it didn't change anything.

If the above example code works in your environment, then I guess there must be something wrong with my test. If this is the case, then I think that this issue is solved. Thanks!

**Comment by gregsmirnov** [comment]
I tried the patch against 3.4.4:
This is an error I get for dev/build

```
Fatal error: Class 'ShortcodeParser' not found in /www/ss34sandbox/framework/core/manifest/ConfigStaticManifest.php(385) : eval()'d code on line 1

Call Stack:
    0.0003     243352   1. {main}() /www/ss34sandbox/framework/cli-script.php:0
    0.0012     301080   2. require_once('/www/ss34sandbox/framework/core/Core.php') /www/ss34sandbox/framework/cli-script.php:64
    5.4652    3092488   3. SS_ConfigStaticManifest->__construct() /www/ss34sandbox/framework/core/Core.php:116
    5.4654    3093480   4. SS_ConfigStaticManifest->regenerate() /www/ss34sandbox/framework/core/manifest/ConfigStaticManifest.php:56
    5.4654    3097384   5. SS_FileFinder->find() /www/ss34sandbox/framework/core/manifest/ConfigStaticManifest.php:109
    6.5314    3984896   6. call_user_func:{/www/ss34sandbox/framework/filesystem/FileFinder.php:158}() /www/ss34sandbox/framework/filesystem/FileFinder.php:158
    6.5314    3984936   7. SS_ConfigStaticManifest->handleFile() /www/ss34sandbox/framework/filesystem/FileFinder.php:158
    6.5332    6018552   8. SS_ConfigStaticManifest_Parser->parse() /www/ss34sandbox/framework/core/manifest/ConfigStaticManifest.php:136
    6.5348    6029408   9. SS_ConfigStaticManifest_Parser->parseStatic() /www/ss34sandbox/framework/core/manifest/ConfigStaticManifest.php:297
    6.5349    6031096  10. eval('static $temp = ShortcodeParser::LEAVE;
return $temp;
') /www/ss34sandbox/framework/core/manifest/ConfigStaticManifest.php:385
```

**Comment by Taitava** [comment]
@gregsmirnov I got that error too at some point yesterday when I was trying the fix, but don't know how I actually solved it. I think I just tried again to remove the `framework` folder (and even the `cms` too just in case, although it shouldn't affect AFAIK) and recopied the patched `framework` folder and the `cms` folder. I don't know why that helped for that particular error (and I know that my method is an overkill anyway). Of course, the original issue hasn't changed, as I mentioned above. But perhaps you can solve that error by just reinstalling the framework and reapplying the patch?

**Comment by gregsmirnov** [comment]
Based on @dhensby work, this should work [918ac95](https://github.com/gregsmirnov/silverstripe-framework/commit/918ac954691672f6b9793e6217da1dc285ad4157). You can download raw patch from [this link](https://github.com/gregsmirnov/silverstripe-framework/commit/918ac954691672f6b9793e6217da1dc285ad4157.patch) to test.

**Comment by Taitava** [comment]
@gregsmirnov Thanks. I applied the patch and can see that it works. Really nice! :)

**Comment by dhensby** [comment]
I'm still trying to get this patch to work consistently, btw. #6619 is now passing but I'd like to optimise a bit more

**Comment by gregsmirnov** [comment]
Updated PR works for me :) I actually test on Security.php, that is included during manifest build.

**Comment by tractorcow** [comment]
Can this be closed now, or is there still work to do, @dhensby ?

**Comment by dhensby** [comment]
This is fixed in 3.x (#6619) and fixed in master by moving to phpparser (#5567)
